ui/src/app/views/breadcrumb/component.tsx

```diff
@@ -0,0 +1,141 @@
+  const {workspace, cohort, conceptSet, urlParams: {ns, wsid, cid, csid, pid, nbName}} = data;
+  const prefix = `/workspaces/${ns}/${wsid}`;
+  switch (type) {
+    case BreadcrumbType.Workspaces:
+      return [
+        {label: 'Workspaces', url: '/workspaces'}
+      ];
+    case BreadcrumbType.Workspace:
+      return [
+        ...getTrail(BreadcrumbType.Workspaces, data),
+        {label: workspace ? workspace.name : '...', url: prefix}
+      ];
+    case BreadcrumbType.WorkspaceEdit:
+      return [
+        ...getTrail(BreadcrumbType.Workspace, data),
+        {label: 'Edit Workspace', url: `${prefix}/edit`}
+      ];
+    case BreadcrumbType.WorkspaceClone:
+      return [
+        ...getTrail(BreadcrumbType.Workspace, data),
+        {label: 'Duplicate Workspace', url: `${prefix}/clone`}
+      ];
+    case BreadcrumbType.Notebook:
+      return [
+        ...getTrail(BreadcrumbType.Workspace, data),
+        {label: 'Notebooks', url: `${prefix}/notebooks`},
+        {
+          label: nbName && decodeURIComponent(nbName).replace(/\.ipynb$/, ''),
+          url: `${prefix}/notebooks/${nbName}`
+        }
+      ];
+    case BreadcrumbType.ConceptSet:
+      return [
+        ...getTrail(BreadcrumbType.Workspace, data),
+        {label: 'Concept Sets', url: `${prefix}/concepts/sets`},
+        {label: conceptSet ? conceptSet.name : '...', url: `${prefix}/concepts/sets/${csid}`}
+      ];
+    case BreadcrumbType.Cohort:
+      return [
+        ...getTrail(BreadcrumbType.Workspace, data),
+        {label: 'Cohorts', url: `${prefix}/cohorts`},
+        {label: cohort ? cohort.name : '...', url: `${prefix}/cohorts/${cid}/review/participants`}
+      ];
+    case BreadcrumbType.Participant:
+      return [
+        ...getTrail(BreadcrumbType.Cohort, data),
+        {label: `Participant ${pid}`, url: `${prefix}/cohorts/${cid}/review/participants/${pid}`}
+      ];
+    case BreadcrumbType.CohortAdd:
+      return [
+        ...getTrail(BreadcrumbType.Workspace, data),
+        {label: 'Build Cohort Criteria', url: `${prefix}/cohorts/build`}
+      ];
+    case BreadcrumbType.Dataset:
+      return [
+        ...getTrail(BreadcrumbType.Workspace, data),
+        {label: 'Dataset', url: `${prefix}/data/datasets`}
+      ];
+    default: return [];
+  }
+};
+
+const BreadcrumbLink = ({href, ...props}) => {
+  return <a
+    href={href}
+    onClick={e => {
+      e.preventDefault();
+      navigateByUrl(href);
+    }}
+    {...props}
+  />;
+};
+
+const Breadcrumb = fp.flow(
+  withCurrentWorkspace(),
+  withCurrentCohort(),
+  withCurrentConceptSet(),
+  withUrlParams(),
+  withRouteConfigData()
+)(({workspace, cohort, conceptSet, urlParams, routeConfigData}) => {
+  const trail = getTrail(
+    routeConfigData.breadcrumb,
+    {urlParams, workspace, conceptSet, cohort}
+  );
+  const first = fp.dropRight(1, trail);
+  const last = fp.last(trail) as any;
```

Comment:
Does this still need an `any` now that you typed it? Or does putting it through lodash drop the type info?

Reply:
Ah, looks like it's no longer needed, thanks.